exist-core/src/main/java/org/exist/xquery/PerformanceStatsImpl.java

```diff
@@ -0,0 +1,421 @@
+            final FunctionStats[] stats = sort();
+            for (final FunctionStats stat : stats) {
+                pw.format("\n%30s %8.3f %8d", stat.qname, stat.executionTime / 1000.0, stat.callCount);
+            }
+            pw.flush();
+            return sw.toString();
+        } catch (final IOException e) {
+            // no-op
+            return "";
+        }
+    }
+
+    private FunctionStats[] sort() {
+        final FunctionStats[] stats = new FunctionStats[functions.size()];
+        int j = 0;
+        for (final FunctionStats next : functions.values()) {
+            stats[j] = next;
+            j++;
+        }
+        Arrays.sort(stats, new CompareByTime());
+        return stats;
+    }
+
+    @Override
+    public void serialize(final MemTreeBuilder builder) {
+        builder.startElement(new QName(XML_ELEMENT_CALLS, XML_NAMESPACE, XML_PREFIX), null);
+        if (isEnabled()) {
+            final AttributesImpl attrs = new AttributesImpl();
+            for (final QueryStats stats : queries.values()) {
+                attrs.clear();
+                attrs.addAttribute("", "source", "source", "CDATA", stats.source);
+                attrs.addAttribute("", "elapsed", "elapsed", "CDATA", Double.toString(stats.executionTime / 1000.0));
+                attrs.addAttribute("", "calls", "calls", "CDATA", Integer.toString(stats.callCount));
+                builder.startElement(new QName("query", XML_NAMESPACE, XML_PREFIX), attrs);
+                builder.endElement();
+            }
+            for (final FunctionStats stats : functions.values()) {
+                attrs.clear();
+                attrs.addAttribute("", "name", "name", "CDATA", stats.qname.getStringValue());
+                attrs.addAttribute("", "elapsed", "elapsed", "CDATA", Double.toString(stats.executionTime / 1000.0));
+                attrs.addAttribute("", "calls", "calls", "CDATA", Integer.toString(stats.callCount));
+                if (stats.source != null) {
+                    attrs.addAttribute("", "source", "source", "CDATA", stats.source);
+                }
+                builder.startElement(new QName("function", XML_NAMESPACE, XML_PREFIX), attrs);
+                builder.endElement();
+            }
+            for (final IndexStats stats : indexStats.values()) {
+                attrs.clear();
+                attrs.addAttribute("", "type", "type", "CDATA", stats.indexType);
+                attrs.addAttribute("", "source", "source", "CDATA", stats.source + " [" + stats.line + ":" +
+                        stats.column + "]");
+                attrs.addAttribute("", "elapsed", "elapsed", "CDATA", Double.toString(stats.executionTime / 1000.0));
+                attrs.addAttribute("", "calls", "calls", "CDATA", Integer.toString(stats.usageCount));
+                attrs.addAttribute("", "optimization-level", "optimization", "CDATA", stats.indexOptimizationLevel.name());
```

Comment:
We could have the old behaviour with
```
attrs.addAttribute("", "optimization", "optimization", "CDATA", stats.indexOptimizationLevel.ordinal());
```

Reply:
You should never use the `ordinal` of an Enum for serialization or deserialization.

Reply:
@line-o I think we agreed at the last meeting to leave the names in favor of numerical values